docs/design/audit-v2.md

```diff
@@ -0,0 +1,79 @@
+
+As the audit observer uses the metainfo loop to iterate through the pointerdb, we're going through all segments and creating reservoirs per unvetted node, and filling the reservoirs with segments to audit.
+The reason for not having reservoirs for all nodes it to minimize memory consumption. We're adding this new method of audit selection is to increase audit frequency for unvetted nodes.
+Once vetted, they should be selected for auditing based on the per stripe (existing selection method), but there is no need to increase the frequency of audits for already-vetted nodes.
+
+We could generate 5 random segments per node, then randomly select some segments per node.
+By choosing nodes out of the reservoirs, we would receive a random sample for auditing.
+
+• We have a reservoir of k items and a stream of n items, where n is an unknown number.
+• Fill the reservoir from [0...k-1] with the first k items in the stream
+• For every item in the stream from index i=k..n-1, pick a random number j=rand(0..i), and if j<k, replace reservoir[j] with stream[i]
+
+## Rationale
+
+An audit observer on the metainfo loop, like every observer on the loop, will always hit every segment on the satellite exactly once, allowing us to get the most accurate possible sample for each node.
+
+While we initially considered integrating the audit system's random node selection process with the existing garbage collection observer, we decided not to do this because the difference in required interval for each observer would mean either too many bloom filters being created unnecessarily or audits occurring too slowly.
+
+An initial idea for implementation was to sort the nodes table for nodes with least amount of audits, then select one node randomly within that low amount of audits.
+However, we decided it may not be necessary to keep track of how many audits per storage node if we're able to randomly select across nodes.
+
+Another approach that we decided not to pursue was the a reverse method of looking up segments or pieces by node ID e.g. a table where each row is a node ID and an encrypted metainfo path.
+Every time a segment is committed or deleted, that table (and every node) gets updated.
+This could simplify the garbage collection process, but complexify upload and download.
+We decided that this would increase database size too significantly to be viable.
+
+## Implementation
+
+1. Add reservoir sampling struct for node auditing.
+2. Create an audit observer that connects to metainfo loop but otherwise uses existing audit code for selection.
+3. Have the audit observer update the reservoir sampling structs.
+3. Make selection biased in favor of unvetted nodes, and select a random segment to audit based on the populated reservoir sampling struct for that node.
+4. Audit the segment selected from part 3, as we are doing already.
+
+## Open issues
```

Comment:
Do we need to fill the reservoir samples every time or can we run it once per day and use the results for multiple audits?

What do we do if the satellite gets restarted? The audit process will start befor the reservoir sampling is finished. How do we handle that?

Do we get problems with the containment mode? Containment mode can deal with only 1 audit at a time (primary key on nodeID). If we run the new audit in parallel the containment mode will fail.

Same interval for both audit jobs? That would allow us to send out up to 2880 audits with up to 80 storage nodes per audit. That is about 230 audits per storage node. Split that halve on both audit jobs? New nodes would get vetted in only 1 day. How do we balance that? (We can increase the vetting requirement to get back to 1 month. However I am not sure if a 50/50 split is the best approach. We could split it 90/10 or what ever. I have no opinion on that. I just want to point it out.

Reply:
> Do we need to fill the reservoir samples every time or can we run it once per day and use the results for multiple audits?

We should be able to use the same reservoir sample for all the segments in the reservoir. So if our reservoir size is 10, we should be able to use that for 10 separate audits for any node without updating it.

> What do we do if the satellite gets restarted? The audit process will start befor the reservoir sampling is finished. How do we handle that?

Not sure if there is a way to get around this issue at the moment. The audit job will just have to wait for the reservoirs to be properly populated. Otherwise all segments don't get an equal chance to be placed in the reservoir.

> Do we get problems with the containment mode? Containment mode can deal with only 1 audit at a time (primary key on nodeID). If we run the new audit in parallel the containment mode will fail.

We can do the same thing that we are doing right now, which is if a node is in containment mode, we re-audit the segment it was put into containment mode for, and do not audit it for the "currently selected" segment.

As far as the intervals for the different audit jobs, I imagine we will make them configurable, so we can adjust them based on our network size and number of vetted/unvetted nodes.

Reply:
> We can do the same thing that we are doing right now, which is if a node is in containment mode, we re-audit the segment it was put into containment mode for, and do not audit it for the "currently selected" segment.

That doesn't work. Lets say both audit jobs are running in parallel. At the point they are starting your node is not in containment mode. At the point they both finish they want to add you to containment mode. The containment mode has a primary key. You can't insert 2 entries. The containment mode is not designed to deal with audit jobs running in parallel. See https://storjlabs.atlassian.net/browse/V3-2071

Reply:
wouldn't the primary key make it so that just one pending audit entry would be inserted? if not, do you know of a way that we could prevent this potential problem?

Reply:
The latest information was that we can replace the existing audit job instead of running two audit jobs in parallel. This means we don't need to worry about the containment mode for the moment and can fix that issue later.